{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "source": [
    "# notebook config (optional)\n",
    "%load_ext lab_black\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%matplotlib inline"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Step 0: MACS2 call peaks from HiChIP data\n",
    "This step is optional if supplied with pre-defined peaks."
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "To call peaks from religation, dangling end, and dump pairs, we use the mapped and paired bam file as input. The pipeline also provide options for including short cis validpairs determined by `local_range` parameters and whether to exlude interchromosomal validpairs (`exclude_interchro`)."
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "source": [
    "from hichip_object.hichip_callpeaks import call_anchors_from_hichip\n",
    "import ray, logging\n",
    "\n",
    "logging.basicConfig(level=logging.INFO, format=\"%(asctime)s: %(message)s\")"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "Fillin the parameters"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "source": [
    "digestion_site = \"(CT[ATCG]AG)|(TTAA)\"  # MiD HiChIP digestion enzymes\n",
    "genome_fa = \"/home/software/mm9.fa\"\n",
    "# for original HiChIP, use merged and paired bam in bwt2/ folder generated by HiC-Pro\n",
    "bam_file = \"/Extension_HDD2/Hanbin/ES_Cell/E14/HiC3_HL/HL28_Smc1_MiDHiChIP_Test/HL28_Smc1_MiDHiChIP_out/bowtie_results/bwt2_multipass/data/data.mergePasses.bam\"\n",
    "# control the PETs to include. Considering includes more validpairs if sequencing depth is low\n",
    "local_range = 1000\n",
    "exclude_interchro = True\n",
    "# macs2 parameters; relax the stringency if sequencing depth is low\n",
    "macs2_qval = 0.01\n",
    "macs2_genome = \"mm\"\n",
    "macs2_path = \"/home/coco/miniconda3/envs/hichip-loop/bin/macs2\"\n",
    "macs2_out_dir = \"/Extension_HDD2/Hanbin/ES_Cell/E14/HiC3_HL/HL28_Smc1_MiDHiChIP_Test/HL28_Smc1_MiDHiChIP_out/\"\n",
    "prefix = \"Smc1_HiChIP_1kb\"\n",
    "\n",
    "mapq = 10\n",
    "nprocs = 30"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "source": [
    "ray.init(num_cpus=nprocs)"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "2021-08-05 16:25:43,823\tINFO services.py:1267 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8266\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "{'node_ip_address': '192.168.0.9',\n",
       " 'raylet_ip_address': '192.168.0.9',\n",
       " 'redis_address': '192.168.0.9:23885',\n",
       " 'object_store_address': '/tmp/ray/session_2021-08-05_16-25-40_062093_38519/sockets/plasma_store',\n",
       " 'raylet_socket_name': '/tmp/ray/session_2021-08-05_16-25-40_062093_38519/sockets/raylet',\n",
       " 'webui_url': '127.0.0.1:8266',\n",
       " 'session_dir': '/tmp/ray/session_2021-08-05_16-25-40_062093_38519',\n",
       " 'metrics_export_port': 59287,\n",
       " 'node_id': 'c9b60367a3a1bca30184f503ffb3ff0fb8b2067daba06b161167b692'}"
      ]
     },
     "metadata": {},
     "execution_count": 4
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "source": [
    "call_anchors_from_hichip(\n",
    "    bam_file,\n",
    "    digestion_site,\n",
    "    genome_fa,\n",
    "    prefix,\n",
    "    macs2_out_dir,\n",
    "    exclude_interchro,\n",
    "    local_range,\n",
    "    macs2_path,\n",
    "    macs2_qval,\n",
    "    macs2_genome,\n",
    "    mapq,\n",
    "    nprocs,\n",
    ")"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "2021-08-05 16:30:54,290: Parsing bam data to Hi-C pairs\n",
      "2021-08-05 16:38:49,593: Removing duplications\n",
      "2021-08-05 16:39:30,293: Processed 71414597 paired PETs. 24626682 paired PETs were kept after duplication removal\n",
      "2021-08-05 16:39:30,296: Flattening to PETs to 100 bp reads\n",
      "2021-08-05 16:41:07,868: 52188049 reads were dumped.\n",
      "2021-08-05 16:41:07,870: Calling peaks by MACS2\n",
      "2021-08-05 16:45:59,342: MACS2 called 143974 peaks at q < 0.01\n"
     ]
    },
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "'/Extension_HDD2/Hanbin/ES_Cell/E14/HiC3_HL/HL28_Smc1_MiDHiChIP_Test/HL28_Smc1_MiDHiChIP_out/Smc1_HiChIP_1kb_MACS2_results/Smc1_HiChIP_1kb_peaks.narrowPeak'"
      ]
     },
     "metadata": {},
     "execution_count": 5
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Step 1: OBJECT call significant interactions"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "source": [
    "# skip this cell if starts from step 0\n",
    "import ray, logging\n",
    "\n",
    "ray.init(num_cpus=30)\n",
    "logging.basicConfig(level=logging.INFO, format=\"%(asctime)s: %(message)s\")"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "2021-08-04 23:03:31,424\tINFO services.py:1267 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8266\u001b[39m\u001b[22m\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "source": [
    "from hichip_object.load_loop_data import process_peak_to_anchor_bins\n",
    "from hichip_object.glm_loop_model import Loop_ZIP"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "Parameters"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "source": [
    "peak_file = \"/Extension_HDD2/Hanbin/ES_Cell/E14/HiC3_HL/HL28_Smc1_MiDHiChIP_Test/HL28_Smc1_MiDHiChIP_out/Smc1_HiChIP_1kb_MACS2_results/Smc1_HiChIP_1kb_peaks.narrowPeak\"\n",
    "bin_size = 2500  # genome is bined and peaks are assigned to the genomic bins\n",
    "chro_size = \"/usr/local/bin/HiC-Pro_2.11.0-beta/annotation/chrom_mm9.sizes\"\n",
    "vp = \"/Extension_HDD2/Hanbin/ES_Cell/E14/HiC3_HL/HL28_Smc1_MiDHiChIP_Test/HL28_Smc1_MiDHiChIP_out/hop_results/data.APASL.mppValidPairs\"\n",
    "inter_range = (5000, 2000000)"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "source": [
    "gb_anchors = process_peak_to_anchor_bins(\n",
    "    peak_file=peak_file,\n",
    "    chro_size=chro_size,\n",
    "    resolution=bin_size,\n",
    ")"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "Load validpair data and mapped to putative anchor pairs"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "source": [
    "loop_zip = Loop_ZIP(\n",
    "    vp,\n",
    "    gb_anchors,\n",
    "    # inter_range = inter_range\n",
    ")"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "2021-08-05 16:48:02,273: Note: NumExpr detected 40 cores but \"NUMEXPR_MAX_THREADS\" not set, so enforcing safe limit of 8.\n",
      "2021-08-05 16:49:24,589: Loading loop PETs\n",
      "2021-08-05 16:51:57,717: Counting anchor depth\n",
      "2021-08-05 16:51:57,940: Joining data\n",
      "2021-08-05 16:55:03,509: 82969 of anchors forms 7663615 putative mid-range loops\n",
      "2021-08-05 16:55:03,570: 839454 observed loops that contains 2257595.0 PETs (avg = 2.6893611800050987)\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "Fit model and calculate p values"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "source": [
    "loop_zip.iteratively_fit_model(disp_glm_summary=False)"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "2021-08-05 16:56:22,912: ========Iter 1========\n",
      "2021-08-05 16:56:22,913: Fitting data to ZeroInflatedPoisson model\n",
      "2021-08-05 16:59:14,504: AIC: 6650659.102784619\n",
      "2021-08-05 16:59:15,164: 51379 interactions were called at qval <= 0.1, count >= 2.6893611800050987\n",
      "2021-08-05 16:59:15,165: ========Iter 2========\n",
      "2021-08-05 16:59:15,695: 9956 high confident interactions were removed from fitting\n",
      "2021-08-05 17:06:00,473: AIC: 5876439.235001802\n",
      "2021-08-05 17:06:02,778: 61760 interactions were called at qval <= 0.1, count >= 2.6893611800050987\n",
      "2021-08-05 17:06:02,779: Sum square changes of E: 82425.70770671748\n",
      "2021-08-05 17:06:02,780: ========Iter 3========\n",
      "2021-08-05 17:06:03,347: 11995 high confident interactions were removed from fitting\n",
      "2021-08-05 17:16:16,835: AIC: 5803013.3594795065\n",
      "2021-08-05 17:16:17,410: 63305 interactions were called at qval <= 0.1, count >= 2.6893611800050987\n",
      "2021-08-05 17:16:17,411: Sum square changes of E: 3001.3492110920974\n",
      "2021-08-05 17:16:17,412: ========Iter 4========\n",
      "2021-08-05 17:16:19,813: 12318 high confident interactions were removed from fitting\n",
      "2021-08-05 17:26:28,205: AIC: 5792430.58363404\n",
      "2021-08-05 17:26:28,734: 63552 interactions were called at qval <= 0.1, count >= 2.6893611800050987\n",
      "2021-08-05 17:26:28,735: Sum square changes of E: 103.34409556160409\n",
      "2021-08-05 17:26:28,736: ========Iter 5========\n",
      "2021-08-05 17:26:29,336: 12373 high confident interactions were removed from fitting\n",
      "2021-08-05 17:36:17,549: AIC: 5790656.16626632\n",
      "2021-08-05 17:36:18,083: 63593 interactions were called at qval <= 0.1, count >= 2.6893611800050987\n",
      "2021-08-05 17:36:18,085: Sum square changes of E: 4.07438349939618\n",
      "2021-08-05 17:36:18,086: ========Iter 6========\n",
      "2021-08-05 17:36:18,678: 12381 high confident interactions were removed from fitting\n",
      "2021-08-05 17:46:33,122: AIC: 5790399.303836828\n",
      "2021-08-05 17:46:33,742: 63600 interactions were called at qval <= 0.1, count >= 2.6893611800050987\n",
      "2021-08-05 17:46:33,743: Sum square changes of E: 0.0642543203200829\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "source": [
    "out = \"loop_zip.self_1kb_peak.bedpe\"\n",
    "loop_zip.write_interaction_statistics(\n",
    "    out,\n",
    ")"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Test overlap with Micro-C loops"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "source": [
    "import pandas as pd\n",
    "from hichip_object.loop_merge import (\n",
    "    combine_two_loop_list,\n",
    "    non_redundant_loops,\n",
    "    significant_loops,\n",
    "    mid_range_loops,\n",
    "    drop_unplaced_chromosome_data,\n",
    "    read_hicuups_loop,\n",
    ")"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "source": [
    "microc_loop_file = \"/Extension_HDD2/Hanbin/ES_Cell/Jm8.N4/Loops/WT_ALL_merge.allValidPairs_5_10kb_res_SCALE_hiccups/merged_loops.bedpe\"\n",
    "microc_loops = non_redundant_loops(\n",
    "    mid_range_loops(\n",
    "        drop_unplaced_chromosome_data(read_hicuups_loop(microc_loop_file))\n",
    "    ).reset_index(drop=True),\n",
    "    10000,\n",
    ")"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "2021-08-05 17:51:41,454: 21030 reduces to 21030\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "source": [
    "out = \"loop_zip.self_1kb_peak.bedpe\"\n",
    "hichip_hightier_loops = non_redundant_loops(\n",
    "    significant_loops(\n",
    "        drop_unplaced_chromosome_data(\n",
    "            pd.read_csv(\n",
    "                out,\n",
    "                sep=\"\\t\",\n",
    "                header=None,\n",
    "                names=[\"chr1\", \"x1\", \"y1\", \"chr2\", \"x2\", \"y2\", \"counts\", \"qval\"],\n",
    "            )\n",
    "        ),\n",
    "        0.05,\n",
    "        3,\n",
    "    ).reset_index(drop=True),\n",
    "    20000,\n",
    ")\n",
    "combine_two_loop_list(\n",
    "    microc_loops, hichip_hightier_loops, 25000, (\"Micro-C\", \"Smc1_MiDHiChIP\")\n",
    ")"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "2021-08-05 17:51:55,584: 50946 reduces to 33732\n",
      "2021-08-05 17:51:55,661: 21030 records in \"Micro-C\" loop set; 50946 records in \"Smc1_MiDHiChIP\" loop set\n",
      "2021-08-05 17:51:55,684: Building Graph\n",
      "2021-08-05 17:52:01,461: Assigning Components\n",
      "2021-08-05 17:52:33,665: Combined to 38632 components (merged loops combined graph of both loop set)\n",
      "2021-08-05 17:52:33,670: Only \"Micro-C\": 8194 component are consist of 8497 \"Micro-C\" non-redundant loops\n",
      "2021-08-05 17:52:33,673: Only \"Smc1_MiDHiChIP\": 19978 component are consist of 20854 \"Smc1_MiDHiChIP\" non-redundant loops\n",
      "2021-08-05 17:52:33,685: 10460 overlaping components are equal to 12533 non-redundant loops for \"Micro-C\" or 12878 for \"Smc1_MiDHiChIP\"\n"
     ]
    }
   ],
   "metadata": {}
  }
 ],
 "metadata": {
  "orig_nbformat": 4,
  "kernelspec": {
   "name": "hichip-loop",
   "display_name": "hichip-py38",
   "language": "python"
  },
  "interpreter": {
   "hash": "c0c3094fc14ca20949d28888a14f8c67ee0128827de5a53627fd5b7bd74589f4"
  },
  "language_info": {
   "name": "python",
   "version": "3.8.10",
   "mimetype": "text/x-python",
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "pygments_lexer": "ipython3",
   "nbconvert_exporter": "python",
   "file_extension": ".py"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}